### ISSUE TITLE
SRFI-37 does not process optinal argument to short option.

### ISSUE THREAD

**Issue by shirok** [shirok/Gauche#703]
From https://sourceforge.net/p/gauche/bugs/75/

Hello.

It looks like Gauche's implementation of SRFI-37 has an issue:

```
gosh> (use srfi-37)
(args-fold '("-I" "the-argument")
                 (list (option '(#\I) #f #t
                               (lambda (opt name arg result)
                                 (and (eqv? name #\I) arg))))
                 (lambda args (error "unrecognized"))
                 (lambda args #f)
                 #f)
#f
```

But the SRFI requires that "the-argument" be recognized as argument to "-I", since "-I" was defined as having an optional argument.

See that, if the option argument is concatenated with the short option name, Gauche correctly recognizes it:

```
(args-fold '("-Ithe-argument")
           (list (option '(#\I) #f #t
                         (lambda (opt name arg result)
                           (and (eqv? name #\I) arg))))
           (lambda args (error "unrecognized"))
           (lambda args #f)
           #f)
```

(This is supposed to work since -I is the last argument)

**Comment by shirok** [comment]
It looks like the original reference implementation has the same issue.

**Comment by shirok** [comment]
Actually, it's not only a short option.  The reference implementation is written so that, if the option takes optional argument, such argument is only recognized when it is given within the same argument.

```
gosh> (args-fold '("--ii" "the-argument")
                 (list (option '("ii") #f #t
                               (lambda (opt name arg result)
                                 arg)))
                 (lambda args (error "unrecognized"))
                 (lambda args #f)
                 #f)
#f
gosh> (args-fold '("--ii=the-argument")
                 (list (option '("ii") #f #t
                               (lambda (opt name arg result)
                                 arg)))
                 (lambda args (error "unrecognized"))
                 (lambda args #f)
                 #f)
"the-argument"
```

If the option takes required argument, it is recognized in a separate argument:

``
gosh> (args-fold '("--ii" "the-argument")
                 (list (option '("ii") #t #f
                               (lambda (opt name arg result)
                                 arg)))
                 (lambda args (error "unrecognized"))
                 (lambda args #f)
                 #f)
"the-argument"
gosh> (args-fold '("--ii=the-argument")
                 (list (option '("ii") #t #f
                               (lambda (opt name arg result)
                                 arg)))
                 (lambda args (error "unrecognized"))
                 (lambda args #f)
                 #f)
"the-argument"
```

So I suspect this could be intentional.  The srfi document actually doesn't clearly mention how "optional" argument should be handled.  We'd better take discussion on the srfi mailing list.

**Comment by jpellegrini** [comment]
Hello; I'm the original issue reporter.

You mean that in this text:
* A short-option-argument is accepted in a separate argument string immediately following a short-option. Examples: `-a foo -bcd bar`
* Also (for historical reasons) a short-option-argument is accepted immediately following the last short-option in a single argument string. Examples: `-afoo -bcdbar`

it's not clear that the argument is of theoptional kind? I understand. Most implementations today seem to mimic this (possibly because they use the reference implementation, or some derived work); only Guile seems different.

**Comment by shirok** [comment]
Yeah, that text only mentions how short option treat its argument.   What matter here isn't about short or long options, but about an "optional" argument (as opposed to a required one).   The reference implementation is consistent that both short and long options only recognizes an argument attached to the option string as the option's "optional" argument.

If srfi text and reference implementation differ, srfi text is definitive and reference implementation should be amended.  I think in this case, it is reasonable that Guile implementation is right, and srfi text needs clarification.

**Comment by shirok** [comment]
For long options, Guile behaves the same as the reference implementation---that is, if the argument is given separately, it isn't recognized as an "optional" argument.   So it's only the short option handling of Guile that differs.  I think if Guile takes the position it handles in the short options, it should do the same for the long options.  Anyway, I'll post to srfi-37 discussion list and see how people think.

```
scheme@(guile-user)> (args-fold '("--ii" "the-argument")
                                (list (option '("ii") #f #t
                                              (lambda (opt name arg result)
                                                arg)))
                                (lambda args (error "unrecognized" args))
                                (lambda args #f)
                                #f)
$4 = #f
scheme@(guile-user)> (args-fold '("--ii=the-argument")
                                (list (option '("ii") #f #t
                                              (lambda (opt name arg result)
                                                arg)))
                                (lambda args (error "unrecognized" args))
                                (lambda args #f)
                                #f)
$5 = "the-argument"
```
